### PR TITLE
Add new model type 'replacement then uninstallation'

### PR DESCRIPTION
**Pull request by ArthurMinfotel**
Combination of both existing model types in one.

### CODE REVIEW

**Comment by stonebuzz** [comment]
Hi @ArthurMinfotel 

If I understand correctly, the new mode allows you to replace an asset with a new one and then uninstall the first?

**Comment by ArthurMinfotel** [comment]
> Hi @ArthurMinfotel
> 
> If I understand correctly, the new mode allows you to replace an asset with a new one and then uninstall the first?

That's exactly it. A typical use case would be : I want to replace computer A by computer B, then put A back in stock.
Right now it would be done by applying a replace type model to A to replace by B, then applying a uninstall type model to A. This is meant to unable doing both action in one go.

**Comment by stonebuzz** [comment]
When the plugin is installed, two models are created by default.

image

A third model would have to be created during the update (or fresh install) to represent this new mode.

**Comment by ArthurMinfotel** [comment]
> When the plugin is installed, two models are created by default.
> 
> image
> 
> A third model would have to be created during the update (or fresh install) to represent this new mode.

It's done.
image

**Review by stonebuzz — APPROVED** [review]

File: inc/model.class.php [hunk 1]
```diff
@@ -85,6 +86,21 @@ public static function getMenuContent()
         return $menu;
     }
 
+    public function prepareInputForAdd($input)
+    {
+        $input = parent::prepareInputForadd($input);
+        if (array_key_exists('types_id', $input) && array_key_exists('replace_method', $input)) {
+            if ($input['types_id'] == self::TYPE_MODEL_REPLACEMENT_UNINSTALL && $input['replace_method'] == PluginUninstallReplace::METHOD_PURGE) {
+                Session::addMessageAfterRedirect(
+                    __("The purge archiving method is not available for this type of model", 'uninstall'),
```
Comment (Rom1-B) [inline inc/model.class.php:95]:
```suggestion
                    __("The purge archiving method is not available for this model type", 'uninstall'),
```

File: inc/model.class.php [hunk 2]
```diff
@@ -129,14 +151,19 @@ public static function getReplacementMethods()
     * Dropdown of method remplacement
     *
     * @param $name   select name
+    * @param $type   int types_id
     * @param $value  default value (default '')
```
Comment (Rom1-B) [inline inc/model.class.php:155]:
```suggestion
    * @param $value  default value (default '')
    * @param $type   int types_id
```

File: inc/model.class.php [hunk 3]
```diff
@@ -673,13 +700,14 @@ public function showFormAction($item)
...
-        } else {
-           // if Replacement is selected
+        if ($this->fields["types_id"] != self::TYPE_MODEL_UNINSTALL) {
+            // if Replacement or Replacement then uninstall is selected
             self::showPartFormRemplacement();
         }
+        if ($this->fields["types_id"] != self::TYPE_MODEL_REPLACEMENT) {
+           // if Uninstall or Replacement then uninstall is selected
+            self::showPartFormUninstall();
+        }
```
Comment (Rom1-B) [inline inc/model.class.php:710]: A switch/case would probably be more suitable

**Review by Rom1-B — CHANGES_REQUESTED** [review]

**Review by Rom1-B — APPROVED** [review]

**Pull request merged**